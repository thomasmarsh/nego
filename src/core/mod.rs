pub mod bitboard;
pub mod coord;
pub mod error;
pub mod game;
pub mod r#move;
pub mod move_tab;
pub mod orientation;
pub mod pieces;
pub mod ray;
pub mod square;
pub mod zobrist;
