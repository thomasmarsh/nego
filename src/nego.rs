mod ai;
mod core;

#[macro_use]
extern crate log;

fn main() -> rustyline::Result<()> {
    pretty_env_logger::init();

    info!("initializing ray LUT");
    core::ray::Rays::build_lut();

    // cli()

    //dem();
    ai::demo_minimax();
    Ok(())
}
