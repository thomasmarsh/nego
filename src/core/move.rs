use crate::core::{
    bitboard::BitBoard,
    coord::{X, Y},
    error::Error,
    move_tab::LUTEntry,
    orientation::Orientation,
    pieces::{PieceId, PieceTypeId},
    square::Square,
};

use serde::Serialize;
use std::fmt;
use std::str::FromStr;

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Serialize)]
pub struct Move(u16);

impl Move {
    #[inline]
    pub fn new(piece: PieceId, entry: LUTEntry) -> Move {
        let p = (piece as u16) & 0b1111;
        let e = (entry.0 as u16) << 4;
        Move(e | p)
    }

    #[inline]
    pub fn get_raw_value(self) -> u16 {
        self.0
    }

    #[inline]
    pub fn get_piece(self) -> PieceId {
        PieceId::from_index(self.0 & 0b1111).unwrap()
    }

    pub fn get_lut_entry(self) -> LUTEntry {
        LUTEntry(self.0.wrapping_shr(4) as usize)
    }

    #[inline]
    pub fn position(&self) -> Square {
        self.get_lut_entry().position()
    }

    #[inline]
    pub fn orientation(&self) -> Orientation {
        self.get_lut_entry().orientation()
    }

    #[inline]
    pub fn mask(&self) -> BitBoard {
        self.get_lut_entry().mask()
    }

    // First set bit (self.position may be empty on kunoji types)
    #[inline]
    pub fn to_square(self) -> Square {
        self.mask().to_square()
    }

    #[inline]
    pub fn gaze(&self) -> BitBoard {
        self.get_lut_entry().gaze()
    }

    pub fn notation(&self) -> String {
        format!(
            "{}:{}{:?}",
            self.get_piece().piece_type_id().notation(),
            self.position().to_string().to_uppercase(),
            self.orientation()
        )
        .to_string()
    }

    pub fn parse(s: &str) -> Result<Move, Error> {
        if s.len() != 8 {
            return Err(Error::InvalidFormat);
        }
        if s.as_bytes()[4] != b':' {
            return Err(Error::InvalidFormat);
        }

        fn parse_piece_type(piece_str: &str) -> Result<PieceTypeId, Error> {
            match piece_str {
                "BOS" => Ok(PieceTypeId::Boss),
                "MAM" => Ok(PieceTypeId::Mame),
                "NOB" => Ok(PieceTypeId::Nobi),
                "KB1" => Ok(PieceTypeId::Koubaku1),
                "KB2" => Ok(PieceTypeId::Koubaku2),
                "KB3" => Ok(PieceTypeId::Koubaku3),
                "KJ1" => Ok(PieceTypeId::Kunoji1),
                "KJ2" => Ok(PieceTypeId::Kunoji2),
                "KJ3" => Ok(PieceTypeId::Kunoji3),
                "KJ4" => Ok(PieceTypeId::Kunoji4),
                _ => Err(Error::InvalidPiece),
            }
        }

        let piece_type = parse_piece_type(&s[0..4])?;
        let x = X::from_str(&s.as_bytes()[5].to_string())?;
        let y = Y::from_str(&s.as_bytes()[6].to_string())?;
        let position = Square::make_square(x, y);
        let orientation = Orientation::from_str(&s.as_bytes()[7].to_string())?;

        let piece = match piece_type {
            PieceTypeId::Boss => PieceId::Boss,
            PieceTypeId::Mame => PieceId::Mame,
            PieceTypeId::Nobi => PieceId::Nobi,
            PieceTypeId::Koubaku1 => PieceId::Koubaku1,
            PieceTypeId::Koubaku2 => PieceId::Koubaku2,
            PieceTypeId::Koubaku3 => PieceId::Koubaku3a, // may need reindexing
            PieceTypeId::Kunoji1 => PieceId::Kunoji1a,   // may need reindexing
            PieceTypeId::Kunoji2 => PieceId::Kunoji2,
            PieceTypeId::Kunoji3 => PieceId::Kunoji3,
            PieceTypeId::Kunoji4 => PieceId::Kunoji4,
        };

        let entry =
            LUTEntry::lookup(piece_type, position, orientation).ok_or(Error::LUTEntryNotFound)?;
        Ok(Move::new(piece, entry))
    }
}

impl fmt::Display for Move {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.notation())
    }
}

impl FromStr for Move {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Move::parse(s)
    }
}

pub trait MoveVisitor {
    fn visit(&mut self, m: Move);

    #[inline]
    fn bailout(&self) -> bool {
        false
    }
}

pub struct HasMoves(pub bool);

impl MoveVisitor for HasMoves {
    #[inline]
    fn visit(&mut self, _: Move) {
        self.0 = true;
    }

    #[inline]
    fn bailout(&self) -> bool {
        self.0
    }
}

pub struct MoveCounter(pub usize);

impl MoveVisitor for MoveCounter {
    #[inline]
    fn visit(&mut self, _: Move) {
        self.0 += 1;
    }
}

pub struct MoveAccumulator(pub Vec<Move>);

impl MoveAccumulator {
    #[inline]
    pub fn new() -> Self {
        Self(Vec::with_capacity(1720 >> 1))
    }
}

impl MoveVisitor for MoveAccumulator {
    #[inline]
    fn visit(&mut self, m: Move) {
        self.0.push(m);
    }
}
