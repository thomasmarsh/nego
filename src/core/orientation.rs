use crate::core::{error::Error, square::Square};

use std::str::FromStr;

#[derive(PartialEq, Clone, Copy, Debug, Eq)]
pub enum Orientation {
    S, // S, 0 (reference orientation of pieces)
    W, // E, 90
    N, // N, 180
    E, // W, 270
}

pub const NUM_ORIENTATIONS: usize = 4;

pub const ALL_ORIENTATIONS: [Orientation; 4] = [
    Orientation::S,
    Orientation::W,
    Orientation::N,
    Orientation::E,
];

impl Orientation {
    #[inline]
    pub fn at_limit(&self, square: Square) -> bool {
        use Orientation::*;
        match self {
            S => square.is_bottom_edge(),
            W => square.is_left_edge(),
            N => square.is_top_edge(),
            E => square.is_right_edge(),
        }
    }

    #[inline]
    pub fn opposite(&self) -> Orientation {
        use Orientation::*;
        match self {
            S => N,
            W => E,
            N => S,
            E => W,
        }
    }

    // if "facing" one direction, what direction is to the right?
    #[inline]
    pub fn right(&self) -> Orientation {
        use Orientation::*;
        match self {
            S => W,
            W => N,
            N => E,
            E => S,
        }
    }

    // if "facing" one direction, what direction is to the left?
    #[inline]
    pub fn left(&self) -> Orientation {
        use Orientation::*;
        match self {
            S => E,
            W => S,
            N => W,
            E => N,
        }
    }

    #[inline]
    pub fn from_index(i: u8) -> Orientation {
        match i {
            0 => Orientation::S,
            1 => Orientation::W,
            2 => Orientation::N,
            3 => Orientation::E,
            _ => unreachable!(),
        }
    }
}

impl FromStr for Orientation {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s.is_empty() {
            return Err(Error::InvalidRank);
        }
        match s.chars().next().unwrap() {
            'S' => Ok(Orientation::S),
            'W' => Ok(Orientation::W),
            'N' => Ok(Orientation::N),
            'E' => Ok(Orientation::E),
            _ => Err(Error::InvalidOrientation),
        }
    }
}
