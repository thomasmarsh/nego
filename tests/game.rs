use nego::core::{
    game::State, move_tab::LUTEntry, orientation::Orientation, pieces::PieceId, r#move::Move,
    square::Square,
};

fn mk_move(piece: PieceId, pos: (u8, u8), orientation: Orientation) -> Move {
    let entry = LUTEntry::lookup(
        piece.piece_type_id(),
        Square::from_indices(pos.0 as usize, pos.1 as usize),
        orientation,
    )
    .unwrap();
    Move::new(piece, entry)
}

fn play_moves(state: &mut State, moves: &[(PieceId, (u8, u8), Orientation)]) {
    moves.iter().for_each(|(a, b, c)| {
        state.place(mk_move(*a, *b, *c));
        state.current = state.current.next();
    });
}

#[test]
fn no_capture_boss() {
    use Orientation::*;
    use PieceId::*;

    let mut state = State::new();
    let first_boss = mk_move(Boss, (0, 0), S);
    state.place(first_boss);
    state.current = state.current.next();
    state.dump();
    let occupied = state.board.black.occupied;

    play_moves(
        &mut state,
        &[
            (Boss, (0, 2), S), // W
            (Mame, (5, 5), S), // B
        ],
    );

    println!("before placement:");
    state.dump();
    state.place(mk_move(Nobi, (2, 0), W)); // W
    println!("after placement:");
    state.dump();

    assert!(state.board.black.occupied & occupied == occupied);
    assert_eq!(state.board.black.move_list[0].get_piece(), Boss);
}

#[test]
fn snapshot() {
    use rand::{seq::SliceRandom, SeedableRng};
    let mut rng = rand::rngs::StdRng::seed_from_u64(123);

    for _ in 0..20 {
        let mut state = State::new();
        let mut moves = Vec::new();
        let mut history = Vec::new();
        loop {
            history.push(state.clone());
            moves.truncate(0);
            state.get_moves(&mut moves);
            if moves.is_empty() {
                break;
            }
            let m = *moves.choose(&mut rng).unwrap();
            state.apply(m);
        }

        insta::assert_debug_snapshot!(history);
    }
}
