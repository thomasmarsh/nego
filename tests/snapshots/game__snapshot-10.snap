---
source: tests/game.rs
expression: history
---
[
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    4095,
                ),
                move_list: [],
                occupied: BitBoard(
                    0,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4095,
                ),
                move_list: [],
                occupied: BitBoard(
                    0,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                0,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 0,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    4094,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                ],
                occupied: BitBoard(
                    12336,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4095,
                ),
                move_list: [],
                occupied: BitBoard(
                    0,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                12336,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 15945533279256367170,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    4094,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                ],
                occupied: BitBoard(
                    12336,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4094,
                ),
                move_list: [
                    Move(
                        128,
                    ),
                ],
                occupied: BitBoard(
                    394752,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                407088,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 796622465954910527,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3070,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                    Move(
                        22410,
                    ),
                ],
                occupied: BitBoard(
                    1614819376,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4094,
                ),
                move_list: [
                    Move(
                        128,
                    ),
                ],
                occupied: BitBoard(
                    394752,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                407088,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 5238794890867998343,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3070,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                    Move(
                        22410,
                    ),
                ],
                occupied: BitBoard(
                    1614819376,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4086,
                ),
                move_list: [
                    Move(
                        128,
                    ),
                    Move(
                        8627,
                    ),
                ],
                occupied: BitBoard(
                    6597070161408,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                407088,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 8192788413904663704,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3068,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                    Move(
                        22410,
                    ),
                    Move(
                        3969,
                    ),
                ],
                occupied: BitBoard(
                    70334296112,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4086,
                ),
                move_list: [
                    Move(
                        128,
                    ),
                    Move(
                        8627,
                    ),
                ],
                occupied: BitBoard(
                    6597070161408,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                407088,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 565064394910915558,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3068,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                    Move(
                        22410,
                    ),
                    Move(
                        3969,
                    ),
                ],
                occupied: BitBoard(
                    70334296112,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4084,
                ),
                move_list: [
                    Move(
                        128,
                    ),
                    Move(
                        8627,
                    ),
                    Move(
                        4289,
                    ),
                ],
                occupied: BitBoard(
                    576467349373584896,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                407088,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 8418808135121228529,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3036,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                    Move(
                        22410,
                    ),
                    Move(
                        3969,
                    ),
                    Move(
                        15749,
                    ),
                ],
                occupied: BitBoard(
                    844495264428080,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4084,
                ),
                move_list: [
                    Move(
                        128,
                    ),
                    Move(
                        8627,
                    ),
                    Move(
                        4289,
                    ),
                ],
                occupied: BitBoard(
                    576467349373584896,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                407088,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 10497115695865769409,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3036,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                    Move(
                        22410,
                    ),
                    Move(
                        3969,
                    ),
                    Move(
                        15749,
                    ),
                ],
                occupied: BitBoard(
                    844495264428080,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4052,
                ),
                move_list: [
                    Move(
                        128,
                    ),
                    Move(
                        8627,
                    ),
                    Move(
                        4289,
                    ),
                    Move(
                        13253,
                    ),
                ],
                occupied: BitBoard(
                    576467353685329408,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                407088,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 7992225589978862835,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3028,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                    Move(
                        22410,
                    ),
                    Move(
                        3969,
                    ),
                    Move(
                        15749,
                    ),
                    Move(
                        8659,
                    ),
                ],
                occupied: BitBoard(
                    870883543494704,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4052,
                ),
                move_list: [
                    Move(
                        128,
                    ),
                    Move(
                        8627,
                    ),
                    Move(
                        4289,
                    ),
                    Move(
                        13253,
                    ),
                ],
                occupied: BitBoard(
                    576467353685329408,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                407088,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 11457015964562856495,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3028,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                    Move(
                        22410,
                    ),
                    Move(
                        3969,
                    ),
                    Move(
                        15749,
                    ),
                    Move(
                        8659,
                    ),
                ],
                occupied: BitBoard(
                    870883543494704,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4036,
                ),
                move_list: [
                    Move(
                        128,
                    ),
                    Move(
                        8627,
                    ),
                    Move(
                        4289,
                    ),
                    Move(
                        13253,
                    ),
                    Move(
                        9940,
                    ),
                ],
                occupied: BitBoard(
                    576467353685378560,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                407088,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 14829355388314793340,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3012,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                    Move(
                        22410,
                    ),
                    Move(
                        3969,
                    ),
                    Move(
                        15749,
                    ),
                    Move(
                        8659,
                    ),
                    Move(
                        10468,
                    ),
                ],
                occupied: BitBoard(
                    14381682425606192,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    4036,
                ),
                move_list: [
                    Move(
                        128,
                    ),
                    Move(
                        8627,
                    ),
                    Move(
                        4289,
                    ),
                    Move(
                        13253,
                    ),
                    Move(
                        9940,
                    ),
                ],
                occupied: BitBoard(
                    576467353685378560,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                407088,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 4568824375099318425,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    3012,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                    Move(
                        22410,
                    ),
                    Move(
                        3969,
                    ),
                    Move(
                        15749,
                    ),
                    Move(
                        8659,
                    ),
                    Move(
                        10468,
                    ),
                ],
                occupied: BitBoard(
                    14381682425606192,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    3012,
                ),
                move_list: [
                    Move(
                        128,
                    ),
                    Move(
                        8627,
                    ),
                    Move(
                        4289,
                    ),
                    Move(
                        13253,
                    ),
                    Move(
                        9940,
                    ),
                    Move(
                        24218,
                    ),
                ],
                occupied: BitBoard(
                    576538547063277056,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                407088,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 9534010396781293837,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    2948,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                    Move(
                        22410,
                    ),
                    Move(
                        3969,
                    ),
                    Move(
                        15749,
                    ),
                    Move(
                        8659,
                    ),
                    Move(
                        10468,
                    ),
                    Move(
                        14374,
                    ),
                ],
                occupied: BitBoard(
                    6931910710066688048,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    3012,
                ),
                move_list: [
                    Move(
                        128,
                    ),
                    Move(
                        8627,
                    ),
                    Move(
                        4289,
                    ),
                    Move(
                        13253,
                    ),
                    Move(
                        9940,
                    ),
                    Move(
                        24218,
                    ),
                ],
                occupied: BitBoard(
                    576538547063277056,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                407088,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 13114670525640136381,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    2948,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                    Move(
                        22410,
                    ),
                    Move(
                        3969,
                    ),
                    Move(
                        15749,
                    ),
                    Move(
                        8659,
                    ),
                    Move(
                        10468,
                    ),
                    Move(
                        14374,
                    ),
                ],
                occupied: BitBoard(
                    6931910710066688048,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    2884,
                ),
                move_list: [
                    Move(
                        128,
                    ),
                    Move(
                        8627,
                    ),
                    Move(
                        4289,
                    ),
                    Move(
                        13253,
                    ),
                    Move(
                        9940,
                    ),
                    Move(
                        24218,
                    ),
                    Move(
                        17543,
                    ),
                ],
                occupied: BitBoard(
                    576538598669993472,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                407088,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 4402527264088983277,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    2820,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                    Move(
                        22410,
                    ),
                    Move(
                        3969,
                    ),
                    Move(
                        15749,
                    ),
                    Move(
                        8659,
                    ),
                    Move(
                        10468,
                    ),
                    Move(
                        14374,
                    ),
                    Move(
                        18567,
                    ),
                ],
                occupied: BitBoard(
                    7223518785938927664,
                ),
                owned: BitBoard(
                    9169073823348162560,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    3012,
                ),
                move_list: [
                    Move(
                        128,
                    ),
                    Move(
                        8627,
                    ),
                    Move(
                        13253,
                    ),
                    Move(
                        9940,
                    ),
                    Move(
                        24218,
                    ),
                    Move(
                        17543,
                    ),
                ],
                occupied: BitBoard(
                    77846366569984,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                407088,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 16128007998819318114,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    2820,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                    Move(
                        22410,
                    ),
                    Move(
                        3969,
                    ),
                    Move(
                        15749,
                    ),
                    Move(
                        8659,
                    ),
                    Move(
                        10468,
                    ),
                    Move(
                        14374,
                    ),
                    Move(
                        18567,
                    ),
                ],
                occupied: BitBoard(
                    7223518785938927664,
                ),
                owned: BitBoard(
                    9169073823348162560,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    2948,
                ),
                move_list: [
                    Move(
                        128,
                    ),
                    Move(
                        8627,
                    ),
                    Move(
                        13253,
                    ),
                    Move(
                        9940,
                    ),
                    Move(
                        24218,
                    ),
                    Move(
                        17543,
                    ),
                    Move(
                        14646,
                    ),
                ],
                occupied: BitBoard(
                    77848522442240,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                407088,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 9893930372654153086,
        capture_flag: false,
    },
    State {
        current: White,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    772,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                    Move(
                        22410,
                    ),
                    Move(
                        3969,
                    ),
                    Move(
                        15749,
                    ),
                    Move(
                        8659,
                    ),
                    Move(
                        10468,
                    ),
                    Move(
                        14374,
                    ),
                    Move(
                        18567,
                    ),
                    Move(
                        26971,
                    ),
                ],
                occupied: BitBoard(
                    7223518785940502576,
                ),
                owned: BitBoard(
                    9169073823348162560,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    2948,
                ),
                move_list: [
                    Move(
                        128,
                    ),
                    Move(
                        8627,
                    ),
                    Move(
                        13253,
                    ),
                    Move(
                        9940,
                    ),
                    Move(
                        24218,
                    ),
                    Move(
                        17543,
                    ),
                    Move(
                        14646,
                    ),
                ],
                occupied: BitBoard(
                    77848522442240,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                407088,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 8797873046657087698,
        capture_flag: false,
    },
    State {
        current: Black,
        board: Board {
            black: PlayerState {
                hand: PieceList(
                    772,
                ),
                move_list: [
                    Move(
                        64,
                    ),
                    Move(
                        22410,
                    ),
                    Move(
                        3969,
                    ),
                    Move(
                        15749,
                    ),
                    Move(
                        8659,
                    ),
                    Move(
                        10468,
                    ),
                    Move(
                        14374,
                    ),
                    Move(
                        18567,
                    ),
                    Move(
                        26971,
                    ),
                ],
                occupied: BitBoard(
                    7223518785940502576,
                ),
                owned: BitBoard(
                    9169073823348162560,
                ),
            },
            white: PlayerState {
                hand: PieceList(
                    2944,
                ),
                move_list: [
                    Move(
                        128,
                    ),
                    Move(
                        8627,
                    ),
                    Move(
                        13253,
                    ),
                    Move(
                        9940,
                    ),
                    Move(
                        24218,
                    ),
                    Move(
                        17543,
                    ),
                    Move(
                        14646,
                    ),
                    Move(
                        4354,
                    ),
                ],
                occupied: BitBoard(
                    77848522442255,
                ),
                owned: BitBoard(
                    0,
                ),
            },
            boss: BitBoard(
                407088,
            ),
            rays: Rays {
                s: BitBoard(
                    0,
                ),
                w: BitBoard(
                    0,
                ),
                n: BitBoard(
                    0,
                ),
                e: BitBoard(
                    0,
                ),
            },
        },
        hash: 12113966931381328910,
        capture_flag: false,
    },
]
